File: Swia/Sources/Swia/types/SkPath.swift
#if canImport(CSkia)
import CSkia
#endif

public class SkPath {
    public static func create() -> SkPath {
        return SkPath(sk_path_new())
    }

    var handle: OpaquePointer

    public init(_ handle: OpaquePointer) {
        self.handle = handle
    }

    public func moveTo(x: Float, y: Float) -> SkPath {
        sk_path_move_to(handle, x, y)
        return self
    }

    public func relativeMoveTo(x: Float, y: Float) -> SkPath {
        sk_path_relative_move_to(handle, x, y)
        return self
    }

    public func lineTo(x: Float, y: Float) -> SkPath {
        sk_path_line_to(handle, x, y)
        return self
    }

    public func relativeLineTo(x: Float, y: Float) -> SkPath {
        sk_path_relative_line_to(handle, x, y)
        return self
    }

    public func addCircle(x: Float, y: Float, r: Float) -> SkPath {
        sk_path_add_circle(handle, x, y, r)
        return self
    }

    public func close() -> SkPath {
        sk_path_close(handle)
        return self
    }

    deinit {
        sk_path_delete(handle)
    }
}

File: cskia/src/sk_path.cc
#include "sk_path.h"
#include "include/core/SkPath.h"
#include "sk_types_priv.h"

sk_path_t* sk_path_new(void) {
    return ToPath(new SkPath());
}

void sk_path_move_to(sk_path_t* path, float x, float y) {
    AsPath(path)->moveTo(x, y);
}

void sk_path_relative_move_to(sk_path_t* path, float x, float y) {
    AsPath(path)->rMoveTo(x, y);
}

void sk_path_line_to(sk_path_t* path, float x, float y) {
    AsPath(path)->lineTo(x, y);
}

void sk_path_relative_line_to(sk_path_t* path, float x, float y) {
    AsPath(path)->rLineTo(x, y);
}

void sk_path_add_circle(sk_path_t* path, float x, float y, float r) {
    AsPath(path)->addCircle(x, y, r, SkPathDirection::kCW);
}

void sk_path_close(sk_path_t* path) {
    AsPath(path)->close();
}

void sk_path_delete(sk_path_t* path) {
    delete AsPath(path);
}
